### PR TITLE
compiler/next: Start parsing and converting internal modules

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
This is for `compiler/next`.

Adjust the `uASTParseFile` function to mimic most of the logic in the
old `parseFile` function. Some things are different (for example, the
new `uAST` builder will handle building an implicit module for us),
but most behaviors should match.

Add a helper function `uASTAttemptToParseMod` which will try to parse
a module into `uAST` and convert it to `AST`. Right now only a subset
of internal modules can be parsed, based on an allow-list that is
baked into the source code. This list should be adjusted over time
as more internal modules can be successfully parsed.

Add comments to the `Module` `uAST` node and adjust its accessor
functions to account for the potential presence of an `Attributes`.

Add a `errors()` iterator to the `ErrorMessage` class.

Adjust iterables I've added to only use a single `using` statement
in order to follow convention.

Attach the `modTag` to the converted `AST` module in `uASTParseFile`.

Adjust converter for `uast::Function` to attach `FLAG_NO_PARENS`.

Add `#defines` which can be used as toggles to try parsing all
internal or standard modules into `uAST`. These may be useful for
further work on this effort.

Clean up calls to `attachSymbolAttributes()` in `convert-uast.cpp`
by making them take a `Decl` instead of an `Attributes` (only `Decl`
can have attributes, anyhow, and all `Decl` are converted into one
or more `Symbol` in the old `AST`).

Reviewed by @mppf. Thanks!

TESTING

- [x] All on `linux64` when `COMM=none` (to make sure `chpl` works)
- [x] All hellos with `--compiler-library-parser`
- [x] All primers with `--compiler-library-parser`

### CODE REVIEW

File: compiler/next/include/chpl/queries/ErrorMessage.h [hunk 1]
```diff
@@ -21,6 +21,7 @@
 #define CHPL_QUERIES_ERRORMESSAGE_H
 
 #include "chpl/queries/Location.h"
+#include "chpl/util/iteration.h"
```
Comment (mppf) [inline compiler/next/include/chpl/queries/ErrorMessage.h:24]: Why is this include here? Did you mean to add a method to ErrorMessage or something?

File: compiler/next/include/chpl/uast/BuilderResult.h [hunk 2]
```diff
@@ -120,8 +120,15 @@ class BuilderResult final {
     return errors_[i];
   }
 
-  // TODO: add something to iterate over the errors e.g.
-  // ErrorListIteratorPair errors() const
+  using ErrorGroup = std::vector<ErrorMessage>;
```
Comment (mppf) [inline compiler/next/include/chpl/uast/BuilderResult.h:123]: It'd be nice to hide that the builder is even using a vector here. I think you could use that by making ErrorGroup an Iterable over vector<ErrorMessage>.

**Review by mppf — APPROVED** [review]
This approach seems fine to me. I expect that as we get everything parsing to uAST, we will no longer have code duplication between uASTParseFile and parseFile. But, what you have here is fine with me in the interim.

File: compiler/next/include/chpl/queries/ErrorMessage.h (hunk 1 above)
Comment (dlongnecke-cray) [reply compiler/next/include/chpl/queries/ErrorMessage.h:24]: The include is here so I can add an a way to iterate over the errors.

Comment (dlongnecke-cray) [reply compiler/next/include/chpl/queries/ErrorMessage.h:24]: Lol, wrong file

File: compiler/next/include/chpl/uast/BuilderResult.h (hunk 2 above)
Comment (dlongnecke-cray) [reply compiler/next/include/chpl/uast/BuilderResult.h:123]: Ok, so something like:

```cpp
using ErrorGroup = Iterable<std::vector<ErrorMessage>>;
using ErrorIterable = ErrorGroup;
```

?

Do you want me to also put `ErrorGroup` into private fields as well?

Comment (mppf) [reply compiler/next/include/chpl/uast/BuilderResult.h:123]: I'm not following at the time why we need both of those but I'd expect we only need one to be public, at lest.

Comment (dlongnecke-cray) [reply compiler/next/include/chpl/uast/BuilderResult.h]: I talked to Andrew and will just define `using ErrorIterable = Iterable<std::vector<thingy>>;` from now on. I think my original idea behind `ErrorGroup` was to make it private, but I'm not sure that works.

**Comment by dlongnecke-cray** [comment]
My expectation is that when we are ready, we will just replace `parseFile` with `uastParseFile` (while renaming the latter).

**Pull request merged**